packages/core/test/application_init_spec.ts

```diff
@@ -28,50 +28,152 @@ import {inject, TestBed, waitForAsync} from '../testing';
+         waitForAsync(inject([ApplicationInitStatus], (status: ApplicationInitStatus) => {
+           // Accessing internal `runInitializers` function of the `ApplicationInitStatus` class
+           // instance for testing purposes to invoke initializer functions.
+           (status as any).runInitializers();
+
+           setTimeout(() => {
+             initFnInvoked = true;
+             reject();
+           });
+
+           expect(status.done).toBe(false);
+           status.donePromise
+               .then(
+                   () => fail('`donePromise.then` should not be invoked when promise is rejected'))
+               .catch(() => {
+                 expect(status.done).toBe(false);
+                 expect(initFnInvoked).toBe(true);
+               });
+         })));
+    });
+
+    describe('with app initializers represented using observables', () => {
+      let subscriber: Subscriber<any>;
+      let observable: Observable<any>;
+      let initFnInvoked = false;
+      beforeEach(() => {
+        observable = new Observable((res) => {
+          subscriber = res;
+        });
+        TestBed.configureTestingModule({
+          providers: [
+            {provide: APP_INITIALIZER, multi: true, useValue: () => observable},
+          ]
+        });
+      });
+
+      it('should update the status once all async observable initializers are completed',
+         waitForAsync(inject([ApplicationInitStatus], (status: ApplicationInitStatus) => {
+           // Accessing internal `runInitializers` function of the `ApplicationInitStatus` class
+           // instance for testing purposes to invoke initializer functions.
+           (status as any).runInitializers();
+
+           setTimeout(() => {
+             initFnInvoked = true;
+             subscriber.complete();
+           });
+
+           expect(status.done).toBe(false);
+           status.donePromise.then(() => {
+             expect(status.done).toBe(true);
+             expect(initFnInvoked).toBe(true);
+           });
+         })));
+
+      it('should update the status once all async observable initializers nexted and completed',
+         waitForAsync(inject([ApplicationInitStatus], (status: ApplicationInitStatus) => {
+           // Accessing internal `runInitializers` function of the `ApplicationInitStatus` class
+           // instance for testing purposes to invoke initializer functions.
+           (status as any).runInitializers();
+
+           subscriber.next('one');
+           subscriber.next('two');
+
+           setTimeout(() => {
+             initFnInvoked = true;
+             subscriber.complete();
+           });
+
+           expect(status.done).toBe(false);
+           status.donePromise.then(() => {
+             expect(status.done).toBe(true);
+             expect(initFnInvoked).toBe(true);
+           });
+         })));
+
+      it('should update the status if all async observable initializers are completed before runInitializers',
```

Comment:
I'd propose to change this into:

```suggestion
      it('should update the status if all async observable initializers are completed synchronously upon subscription',
```

The ordering with respect to `runInitializers` is an implementation detail of the test to achieve synchronous completion, by ensuring the Observable immediately completes.

Reply:
@vthinkxie just a quick update: we've discussed this with Joost and I plan to push a commit to this PR with tests refactoring that would test `ApplicationInitStatus` directly to avoid TestBed side-effects of invoking `runInitializers`. I will leave a comment here once it's done. Thank you.